File: org.dcs.core/src/main/java/org/dcs/core/module/flow/ModuleFactoryServiceImpl.java
package org.dcs.core.module.flow;

import java.util.Map;
import java.util.UUID;
import java.util.concurrent.ConcurrentHashMap;

import javax.enterprise.inject.Default;
import javax.inject.Inject;

import org.dcs.api.service.ModuleFactoryService;
import org.dcs.api.service.RESTException;
import org.ops4j.pax.cdi.api.OsgiServiceProvider;
import org.ops4j.pax.cdi.api.Properties;
import org.ops4j.pax.cdi.api.Property;
import org.osgi.framework.BundleContext;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

@OsgiServiceProvider
@Properties({
    @Property(name = "service.exported.interfaces", value = "*"),
    @Property(name = "service.exported.configs", value = "org.apache.cxf.ws")
})
@Default
public class ModuleFactoryServiceImpl implements ModuleFactoryService {

	private static final Logger logger = LoggerFactory.getLogger(ModuleFactoryServiceImpl.class);

	private ConcurrentHashMap<String, FlowModule> flowModuleMap = new ConcurrentHashMap<>();

	@Inject
	private BundleContext bundleContext;

	/* (non-Javadoc)
	 * @see org.dcs.core.module.flow.ModuleFactoryService#createFlowModule(java.lang.String)
	 */
	@Override
	public String createFlowModule(String className) {
		String randomUUID = UUID.randomUUID().toString();

		try {
			Class<?> clazz = Class.forName(className);			 			
			Object obj = clazz.newInstance();

			if(obj instanceof FlowModule ) {
				FlowModule fm = (FlowModule)obj;
				fm.init(bundleContext);
				flowModuleMap.put(randomUUID, fm);
			} else {
				logger.warn("Given classname " + className + " is not of type FlowModule");
				return null;
			}
			return randomUUID;
		} catch (ClassNotFoundException | InstantiationException | IllegalAccessException e) {
			logger.warn("Error initialising class " + className);
			return null;
		}
	}


	public FlowModule getModule(String moduleUUID) {
		return flowModuleMap.get(moduleUUID);
	}


	/* (non-Javadoc)
	 * @see org.dcs.core.module.flow.ModuleFactoryService#getPropertyDescriptors(java.util.UUID)
	 */
	@Override
	public Map<String, java.util.Properties> getPropertyDescriptors(String moduleUUID) {
		return getModule(moduleUUID).getPropertyDescriptors();
	}

	/* (non-Javadoc)
	 * @see org.dcs.core.module.flow.ModuleFactoryService#getRelationships(java.util.UUID)
	 */
	@Override
	public Map<String, java.util.Properties> getRelationships(String moduleUUID) {
		return getModule(moduleUUID).getRelationships();
	}

	/* (non-Javadoc)
	 * @see org.dcs.core.module.flow.ModuleFactoryService#schedule(java.util.UUID)
	 */
	@Override
	public boolean schedule(String moduleUUID) {
		getModule(moduleUUID).schedule();
		return true;
	}

	/* (non-Javadoc)
	 * @see org.dcs.core.module.flow.ModuleFactoryService#trigger(java.util.UUID, org.apache.nifi.processor.ProcessContext)
	 */
	@Override
	public Object trigger(String moduleUUID,  Map<String, java.util.Properties> properties) throws RESTException {		
		return getModule(moduleUUID).trigger(properties);		
	}

	/* (non-Javadoc)
	 * @see org.dcs.core.module.flow.ModuleFactoryService#unschedule(java.util.UUID)
	 */
	@Override
	public boolean unschedule(String moduleUUID) {
		getModule(moduleUUID).unschedule();
		return true;
	}

	/* (non-Javadoc)
	 * @see org.dcs.core.module.flow.ModuleFactoryService#stop(java.util.UUID)
	 */
	@Override
	public boolean stop(String moduleUUID) {
		getModule(moduleUUID).stop();
		return true;
	}

	/* (non-Javadoc)
	 * @see org.dcs.core.module.flow.ModuleFactoryService#shutdown(java.util.UUID)
	 */
	@Override
	public boolean shutdown(String moduleUUID) {
		getModule(moduleUUID).shutdown();
		return true;
	}

	/* (non-Javadoc)
	 * @see org.dcs.core.module.flow.ModuleFactoryService#remove(java.util.UUID)
	 */
	@Override
	public boolean remove(String moduleUUID) {
		try {
			getModule(moduleUUID).remove();
		} finally {
			flowModuleMap.remove(moduleUUID);
		}
		return true;
	}
}

File: org.dcs.core/src/main/scala/org/dcs/core/module/flow/ModuleFactoryServiceImpl.scala
package org.dcs.core.module.flow

import java.util.{UUID, Map => JavaMap}
import javax.enterprise.inject.Default
import javax.inject.Inject

import org.dcs.api.error.{ErrorConstants, RESTException}
import org.dcs.api.service.ModuleFactoryService
import org.ops4j.pax.cdi.api.{OsgiService, OsgiServiceProvider, Properties, Property}
import org.osgi.framework.BundleContext
import org.slf4j.{Logger, LoggerFactory}

import scala.collection.mutable.{Map => MutableMap}

@OsgiServiceProvider
@Properties(Array(
  new Property(name = "service.exported.interfaces", value = "*"),
  new Property(name = "service.exported.configs", value = "org.apache.cxf.ws")))
@Default
class ModuleFactoryServiceImpl extends ModuleFactoryService {

  val logger: Logger = LoggerFactory.getLogger(classOf[ModuleFactoryServiceImpl])

  var flowModuleMap: Map[String, FlowModule] = Map()

  @Inject
  var bundleContext: BundleContext = _

  override def createFlowModule(className: String): String = {

    val randomUUID: String = UUID.randomUUID().toString()
    val clazz = Class.forName(className)
    val ref = clazz.newInstance

    ref match {
      case ref: FlowModule => {
        ref.init(bundleContext);
        flowModuleMap = flowModuleMap + (randomUUID -> ref);
      }
      case s: String => {
        logger.warn("Given classname " + className + " is not of type FlowModule");
        throw new RESTException(ErrorConstants.DCS001)
      }
    }
    randomUUID;
  }

  def getModule(moduleUUID: String): FlowModule = {
    flowModuleMap.getOrElse(moduleUUID, null)
  }

  override def getPropertyDescriptors(moduleUUID: String): JavaMap[String, JavaMap[String, String]] = {
    getModule(moduleUUID).getPropertyDescriptors();
  }

  override def getRelationships(moduleUUID: String): JavaMap[String, JavaMap[String, String]] = {
    getModule(moduleUUID).getRelationships();
  }

  override def schedule(moduleUUID: String): Boolean = {
    getModule(moduleUUID).schedule();
    true;
  }

  override def trigger(moduleUUID: String, properties: JavaMap[String, String]): Array[Byte] = {
    getModule(moduleUUID).trigger(properties);
  }

  override def unschedule(moduleUUID: String): Boolean = {
    getModule(moduleUUID).unschedule();
    true;
  }

  override def stop(moduleUUID: String): Boolean = {
    getModule(moduleUUID).stop();
    true;
  }

  override def shutdown(moduleUUID: String): Boolean = {
    getModule(moduleUUID).shutdown();
    true;
  }

  override def remove(moduleUUID: String): Boolean = {
    try {
      getModule(moduleUUID).remove();
    } finally {
      flowModuleMap -= (moduleUUID);
    }
    return true;
  }
}
